src/plugins/ds_mongo.c

```diff
@@ -0,0 +1,3126 @@
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <grp.h>
+#include <pthread.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <libyang/libyang.h>
+
+#include "ds_mongo_config.h"
+#include "plugins_datastore.h"
+#include "sysrepo.h"
+
+#define plugin_name "MONGO DS"
+
+#define WRAPERR(err, type, func, message) srplg_log_errinfo(err, plugin_name, NULL, type, func " failed on %d in " __FILE__ " [%s].", __LINE__, message);
+
+typedef struct mongo_data_s {
+    /* client that connects to the database and
+            manages changes on the data */
+    mongoc_client_t *client;
+    /* specific database that is currently being
+            managed (startup, running, ...) */
+    mongoc_database_t *datastore;
+    /* specific collection that is currently being
+            managed (contains data of a specific YANG module) */
+    mongoc_collection_t *module;
+    /* allocated module name of the module */
+    char *module_name;
+} mongo_data_t;
+
+typedef struct mongo_plg_conn_data_s {
+    /* pool of clients that connect to the database
+            so that multithreading is supported, pool is the used
+            pool of clients */
+    mongoc_client_pool_t *pool;
+} mongo_plg_conn_data_t;
+
+/* specific data for MongoDB */
+typedef struct mongo_plg_data_s {
+    /* global variable that checks whether mongoc_init() was called */
+    int is_mongoc_initialized;
+    /* mutex */
+    pthread_mutex_t lock;
+} mongo_plg_data_t;
+
+mongo_plg_data_t plugin_data = {0};
+
+/* helper structure for storing bson data */
+struct mongo_diff_inner_data {
+    bson_t **docs;
+    uint32_t size;
+    uint32_t idx;
+};
+
+/* helper structure for storing data to be committed */
+struct mongo_diff_data {
+/* arrays for elements of the YANG module (without user-ordered elements) */
+/* array for storing nodes with create operation */
+    struct mongo_diff_inner_data cre;
+
+/* array for storing nodes with delete operation */
+    struct mongo_diff_inner_data del;
+
+/* array for storing nodes with replace operation */
+    struct mongo_diff_inner_data rep;
+
+/* array for storing selectors for replace operation */
+    struct mongo_diff_inner_data rep_keys;
+};
+
+static void
+terminate(void)
+{
+    mongoc_cleanup();
+}
+
+static const char *
+ds2str(sr_datastore_t ds)
+{
+    switch (ds) {
+    case SR_DS_STARTUP:
+        return "sr_startup";
+    case SR_DS_RUNNING:
+        return "sr_running";
+    case SR_DS_CANDIDATE:
+        return "sr_candidate";
+    case SR_DS_OPERATIONAL:
+        return "sr_operational";
+    case SR_DS_FACTORY_DEFAULT:
+        return "sr_factory-default";
+    default:
+        return NULL;
+    }
+}
+
+/* try a general command in order to establish whether authentication is required */
+static sr_error_info_t *
+srpds_try_command(mongoc_client_t *client, int *auth_prob)
```

Comment:
No doxygen function comment and if the function is used only for checking authentication, it should be called that way.

Reply:
done